### PR TITLE
[REG-2230] Operate on BotSegmentLists so we can do validations there

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
- Changes the container to use lists of BotSegmentLists instead of lists of botsegments so that we can do validations at the BotSegmentList level

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs [hunk 1]
```diff
@@ -108,6 +110,21 @@ private void ProcessBotSegments()
                         ProcessBotSegmentAction(nextBotSegment, transformStatuses, entityStatuses);
                     }
 
+                    // Run (and potentially end) the validations for this segment
+                    nextBotSegment.ProcessValidation();
+
+                    // Also run the validations for the botsegmentList if they exist
+                    foreach (var validation in _botSegmentListValidations)
+                    {
+                        validation.ProcessValidation(nextBotSegment.Replay_SegmentNumber);
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs:119]: So this is passing down the segment number, but the validation is not really part of that segment, it is part of that segment list as a whole... do we think this could confuse the user? This is why I used -1 for sequences, and then in the logs that gets converted to "SEQUENCE"

**Review by vontell — COMMENTED** [review]
I'll check out the branch and see if it aligns with what I expect, but in the meantime just one comment around the logging of work on validations that are not technically running as part of a particular segment

**Review by vontell — CHANGES_REQUESTED** [review]
(forgot to mark as request changes)

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs (hunk 1 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs:119]: i want to see the active segment in the logs as this runs so I can see what segment I was on as the validation was running

Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs:119]: they are absolutely being evaluated during that currently active segment, and should log as such

Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs:119]: Sounds good to me 👍 I can get behind that logic

**Review by vontell — APPROVED** [review]

**Comment by RG-nAmKcAz** [comment]
@vontell  feel free to merge (will go into your branch) ... as soon as you're done testing / happy with it

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs [hunk 2]
```diff
@@ -192,7 +209,7 @@ private void ProcessBotSegments()
                         {
                             if (_nextBotSegments.Count < 2)
                             {
-                                var next = _dataPlaybackContainer.DequeueBotSegment();
+                                var next = _dataPlaybackContainer.DequeueBotSegment(out _botSegmentListValidations);
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs:212]: @RG-nAmKcAz quick question on this actually regarding the lookahead logic. Does this mean that the currently running validation will be lost and replaced? I am not full understanding this code yet but I wanted to check if this means that validations that are currently running and not yet finished may then be running forever, since now we only operate on these new validations. Maybe we don't need to worry about that because we run StopValidations on the data container itself which does hold onto the reference...

Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentsPlaybackController.cs:212]: i'll huddle with you on this

Stop I think we're fine

but you correctly point out that we may not be considering all the proper validations... I want to discuss if we get rid of this look ahead logic all together or not

**Comment by vontell** [comment]
Closing this PR as it has now been merged into the other feature branch from #375
